rbd-target-api.py

```diff
@@ -798,6 +799,7 @@ def _disk(image_id):
         # A put is for either a create or a resize
         # put('http://localhost:5000/api/disk/rbd.ansible3',
         #     data={'pool': 'rbd','size': '3G','owner':'ceph-1'})
+        LUN.clear_tcm_cache()
```

Comment:
The cache clear will force us to re-read the bs state and avoid the problem, but if we clear it for every disk/lun operation then it can get expensive re-creating the cache for each operation. If you did all the disk creations through gwcli and did something like:

for i > 256++
    disk add

the gluster devs were hitting 1 sec for the equivalent of the first disk addition, and by the time it got to the 256th one it would take almost 30 seconds to recreate the cache.

In the logs I think they are hitting the case where ansible and gwcli are both used. ansible created the disk in kernel and rbd-target-api just did not get updated so its bs_cache was out of sync. We have a similar issue with the config object and were using rbd-target-api:ConfigWatcher() to try and keep that updated. We could do the cache invalidate call in there so for the gwcli-only use case we would not have to invalidate it every time. We would still have a race windows and hit a similar case where the bs cache is out of sync if something like gluster block or targetcli is used directly for some other target (do we support that?).

Or maybe do we just want to modify rtslib so if _index == None but it is in configfs, we add the SO to the cache instead of failing?

Reply:
@mikechristie I was thinking that it's a bug that definitely should get eventually fixed in ```rtslib_fb``` to just query the configfs if the index lookup fails -- this is just an immediate workaround for the LUN creation/update case. In typical operation, nobody will be using gwcli to create hundreds of LUNs (unlike Gluster).

The idea of ansible vs gwcli updates causing the issue is certainly plausible. My guess was just that it was a race between the startup of rbd-target-gw creating the LUNs on reboot and the -api building its cache for the first time.

Reply:
Darn. I made the rtslib fix but realized we have to do invalidate/clear cache. The problem is that after multiple additions/deletions the name to index mapping will be completely different.

Will make a patch to rbd-target-gw similar to this one,